hivemind/utils/mpfuture.py

```diff
@@ -2,171 +2,262 @@
+TERMINAL_STATES = {base.FINISHED, base.CANCELLED, base.CANCELLED_AND_NOTIFIED}
 
+try:
+    from concurrent.futures import InvalidStateError
+except ImportError:
+    # Python 3.7 doesn't raise concurrent.futures.InvalidStateError for repeating set_result/set_exception calls and
+    # doesn't even define this error. In this module, we simulate the Python 3.8+ behavior,
+    # defining and raising this error if necessary.
+    class InvalidStateError(Exception):
+        """Raised when attempting to change state of a future in a terminal state (e.g. finished)"""
 
-class FutureStateError(RuntimeError):
-    """Raised when attempting to change state of a future in a terminal state (e.g. finished)"""
-    pass
+
+class UpdateType(Enum):
+    RESULT = auto()
+    EXCEPTION = auto()
+    CANCEL = auto()
 
 
 class MPFuture(base.Future, Generic[ResultType]):
-    """ Multiprocessing version of concurrent.futures.Future. Can also be awaited like asyncio.Future """
+    """
+    A version of concurrent.futures.Future / asyncio.Future that can be fulfilled from a separate process.
+    Any process can access future status and set the result / exception and check for state.
+    However, only the original process (i.e. the process that created the future) can await the result or exception.
+
+    :param use_lock: if True, operations with MPFuture use a global lock to prevent concurrent writes to the same pipe;
+      If set to False, writing to this future ignores global lock, slightly improving performance, but making user
+      responsible for avoiding concurrent set_result / set_exception calls to futures with the same process of origin.
+    :param loop: if specified, overrides default asyncio event loop for the purpose of awaiting MPFuture
+
+    :note: This is an internal primitive that is not guaranteed to work outside of hivemind applications.
+     More specifically, there are two known limitations:
+       - MPFuture works between processes created through inheritance (e.g. fork), *not* for independent processes
+       - MPFuture is deterministic if only one process can call set_result/set_exception/set_running_or_notify_cancel
+         and only the origin process can call result/exception/cancel.
+    """
+    _initialization_lock = mp.Lock()  # global lock that prevents simultaneous initialization of two processes
+    _update_lock = mp.Lock()  # global lock that prevents simultaneous writing to the same pipe
+    _global_sender_pipe: Optional[PipeEnd] = None  # a pipe that is used to send results/exceptions to this process
+    _pipe_waiter_thread: Optional[threading.Thread] = None  # process-specific thread that receives results/exceptions
+    _active_futures: Optional[Dict[UID, MPFuture]] = None  # pending or running futures originated from current process
+    _active_pid: Optional[PID] = None  # pid of currently active process; used to handle forks natively
```

Comment:
Is it a really good practice, to use `None` as a default value, if it's never used, or even checked for?
At least for `_active_futures` it seems really simple to have `{}` as a default.

Reply:
Though mutable defaults are generally frowned upon in function arguments, perhaps in that case it's ok, since we use the global class field anyway

And for non-mutable ones, maybe just a type annotation would suffice (unless we need to explicitly check that is was not initialized)

Reply:
We have a similar pattern/issue in other parts of the repo, let's stick to None for now and discuss global issue on the nearest meeting